### PR TITLE
Add support for multiple interceptors.

### LINKED ISSUE #249: Support multiple interceptors
# Expected Behavior

A trigger can specify multiple interceptors. 

# Actual Behavior

A trigger can only have a single interceptor.

# Additional Info

Would be needed if we go with Option 3 from the [Filters design doc](https://docs.google.com/document/d/1c-cGLdRb4LuH2_rUn50cIhHPabAoj4iQvw49L4T2nSU/edit#) (#49)

### PR DESCRIPTION
**Pull request by wlynch**
# Changes

* Adds support for multiple interceptors to be chained together.
* Adds new `interceptors` field to EventListenerTrigger, converts
  now deprecated `interceptor` field to `interceptors` automatically.
* Separates interceptor processing into its own func for easier testing.
* Adds basic sink level unit testing for interceptors.

Fixes #249 

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/triggers/blob/master/CONTRIBUTING.md) for more details._

# Release Notes

```
interceptor field is now deprecated in favor of interceptors. Old field is backwards compatible for the time being, but will be removed in a future version.
```

### CODE REVIEW

File: pkg/apis/triggers/v1alpha1/event_listener_validation_test.go [hunk 1]
```diff
@@ -153,6 +153,31 @@ func Test_EventListenerValidate(t *testing.T) {
 				),
 				bldr.EventListenerTrigger("tb", "tt", "v1alpha1"))),
 		raiseErr: false,
+	}, {
+		name: "Valid EventListener Multiple Interceptor",
```
Comment (dibyom) [inline pkg/apis/triggers/v1alpha1/event_listener_validation_test.go]: Would it be valid to have multiple interceptor but of different types e.g. Github + Webhook. Going through the current validation code, it seems like that would result in an error? https://github.com/tektoncd/triggers/blob/master/pkg/apis/triggers/v1alpha1/event_listener_validation.go#L102

It seems like multiple webhooks interceptors should be allowed. Not sure if there is a valid use case for having two Github/Gitlab interceptors though

File: pkg/apis/triggers/v1alpha1/event_listener_defaults.go [hunk 2]
```diff
@@ -10,17 +10,34 @@ func (el *EventListener) SetDefaults(ctx context.Context) {
...
+			upgradeBinding(t)
+			upgradeInterceptor(t)
 		}
 	}
 }
+
+func upgradeBinding(t *EventListenerTrigger) {
+	if t.DeprecatedBinding != nil {
+		if len(t.Bindings) > 0 {
+			// Do nothing since it will be a Validaiton Error.
```
Comment (bobcatfish) [inline pkg/apis/triggers/v1alpha1/event_listener_defaults.go]: s/Validaiton/Validation/

(below also!

File: pkg/sink/sink.go [hunk 3]
```diff
@@ -150,6 +112,63 @@ func (r Sink) HandleEvent(response http.ResponseWriter, request *http.Request) {
...
+	params, err := template.ResolveParams(rt.TriggerBindings, finalPayload, request.Header, rt.TriggerTemplate.Spec.Params)
+	if err != nil {
+		log.Error(err)
+		return err
+	}
+	log.Info("params: %+v", params)
+	resources := template.ResolveResources(rt.TriggerTemplate, params)
+	if err := r.createResources(resources, t.Name, eventID); err != nil {
+		log.Error(err)
+		return err
```
Comment (bobcatfish) [inline pkg/sink/sink.go]: to make it a bit easier to trace through errors, could we wrap the error with a bit of context here? e.g. fmt.Errorf("something useful: %w", err)

File: pkg/sink/sink.go [hunk 4]
```diff
@@ -150,6 +112,63 @@ func (r Sink) HandleEvent(response http.ResponseWriter, request *http.Request) {
...
+	log := eventLog.With(zap.String(triggersv1.TriggerLabelKey, t.Name))
+
+	finalPayload, err := r.executeInterceptors(t, request, event, eventID, log)
+	if err != nil {
+		return err
+	}
+
+	rt, err := template.ResolveTrigger(*t,
+		r.TriggersClient.TektonV1alpha1().TriggerBindings(r.EventListenerNamespace).Get,
+		r.TriggersClient.TektonV1alpha1().TriggerTemplates(r.EventListenerNamespace).Get)
```
Comment (bobcatfish) [inline pkg/sink/sink.go]: Thanks for refactoring these functions out of the big one above!!

if we pass these clients into this funciton, then we can move the new functions into another package and have unit tests for them 🙏

File: pkg/sink/sink.go [hunk 5]
```diff
@@ -150,6 +112,63 @@ func (r Sink) HandleEvent(response http.ResponseWriter, request *http.Request) {
...
+		log.Error(err)
+		return err
+	}
+
+	params, err := template.ResolveParams(rt.TriggerBindings, finalPayload, request.Header, rt.TriggerTemplate.Spec.Params)
+	if err != nil {
+		log.Error(err)
+		return err
+	}
+	log.Info("params: %+v", params)
```
Comment (bobcatfish) [inline pkg/sink/sink.go]: if we want to keep this log message can we add a bit more info? (also i think we need Infof?)

File: pkg/sink/sink.go [hunk 6]
```diff
@@ -150,6 +112,63 @@ func (r Sink) HandleEvent(response http.ResponseWriter, request *http.Request) {
...
+		switch {
+		case i.Webhook != nil:
+			interceptor = webhook.NewInterceptor(i.Webhook, r.HTTPClient, r.EventListenerNamespace, log)
+		case i.Github != nil:
+			interceptor = github.NewInterceptor(i.Github, r.KubeClientSet, r.EventListenerNamespace, log)
+		case i.Gitlab != nil:
+			interceptor = gitlab.NewInterceptor(i.Gitlab, r.KubeClientSet, r.EventListenerNamespace, log)
+		}
+
+		// TODO(https://github.com/tektoncd/triggers/issues/269): Chain interceptors
```
Comment (bobcatfish) [inline pkg/sink/sink.go]: nit: usually we do just TODO(#269)

File: pkg/sink/sink_test.go [hunk 7]
```diff
@@ -478,3 +481,59 @@ func Test_addLabels(t *testing.T) {
...
+	var got map[string]int
+	if err := json.Unmarshal(resp, &got); err != nil {
+		t.Fatalf("json.Unmarshal: %v", err)
+	}
+	want := map[string]int{"i": 2}
+	if diff := cmp.Diff(want, got); diff != "" {
+		t.Errorf("-want +got: %s", diff)
+	}
+
+}
```
Comment (bobcatfish) [inline pkg/sink/sink_test.go]: nice test!

File: pkg/sink/sink_test.go [hunk 8]
```diff
@@ -478,3 +481,59 @@ func Test_addLabels(t *testing.T) {
...
+
+func (f *fakeHTTPInterceptor) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	data := map[string]interface{}{
+		"i": atomic.AddInt32(&f.i, 1),
+	}
+	_ = json.NewEncoder(w).Encode(data)
+}
+
+func TestExecuteInterceptor(t *testing.T) {
```
Comment (bobcatfish) [inline pkg/sink/sink_test.go]: this test is super cool: might be worth adding a docstring explaining how it works? e.g. uses 2 interceptors which increment an int so you can be sure they both are executed, and in order, etc.

File: pkg/apis/triggers/v1alpha1/event_listener_validation_test.go [hunk 9]
```diff
@@ -153,6 +153,31 @@ func Test_EventListenerValidate(t *testing.T) {
...
+			Spec: v1alpha1.EventListenerSpec{
+				Triggers: []v1alpha1.EventListenerTrigger{{
+					Bindings:              []*v1alpha1.EventListenerBinding{{Name: "tb"}},
+					Template:              v1alpha1.EventListenerTemplate{Name: "tt"},
+					DeprecatedInterceptor: &v1alpha1.EventInterceptor{},
+					Interceptors:          []*v1alpha1.EventInterceptor{{}, {}},
+				}},
+			},
+		},
+		raiseErr: true,
```
Comment (bobcatfish) [inline pkg/apis/triggers/v1alpha1/event_listener_validation_test.go]: if you feel motivated, i find it much easier to maintain when there is one test for success cases and a separate test for error cases - much easier to see what is expected to be valid and what isnt

File: pkg/sink/sink_test.go (hunk 7 above)
Comment (bobcatfish) [reply pkg/sink/sink_test.go]: are there any other test cases that we might want to include as well? e.g. if one interceptor fails, should the next be invoked?

Comment (bobcatfish) [inline pkg/sink/sink_test.go]: nice test!!

Comment (bobcatfish) [reply pkg/sink/sink_test.go]: are there other test cases we might want to add as well, e.g. what if one interceptor fails, should the next be run?

Comment (bobcatfish) [reply pkg/sink/sink_test.go]: that particular case could maybe also be handled in a unit test for `executeInterceptors` :D

File: pkg/sink/sink_test.go (hunk 8 above)
Comment (bobcatfish) [inline pkg/sink/sink_test.go]: might be useful to include a brief docstring explaining how this test worked - after reading through it it seems super cool ✨ but had to read a bunch before understanding it (that the interceptor increments an int etc)

File: pkg/apis/triggers/v1alpha1/event_listener_defaults.go (hunk 2 above)
Comment (wlynch) [reply pkg/apis/triggers/v1alpha1/event_listener_defaults.go]: Done.

File: pkg/sink/sink_test.go (hunk 8 above)
Comment (wlynch) [reply pkg/sink/sink_test.go]: Done.

Comment (wlynch) [reply pkg/sink/sink_test.go]: Done.

File: pkg/interceptors/webhook/interceptors.go [hunk 10]
```diff
@@ -47,6 +47,7 @@ func createOutgoingRequest(ctx context.Context, original *http.Request, url *url
 	r := original.Clone(ctx)
 	r.RequestURI = "" // RequestURI cannot be set in outgoing requests
 	r.URL = url
+	r.Host = url.Host
```
Comment (dibyom) [inline pkg/interceptors/webhook/interceptors.go]: :+1: TIL 

>// For incoming requests, the Host header is promoted to the
    // Request.Host field and removed from the Header map.

**Comment by dibyom** [comment]
Ohh, one more thing (really a question) that I just realized :sweat_smile: It seems like we are planning to change what body each interceptor in receives in #269 This seems like a breaking change for a user using multiple interceptors so my question is should we attempt #269 as part of 0.2? 
While this is definite scope creep, it might be worth reducing the number of breaking changes?

**Comment by wlynch** [comment]
> Ohh, one more thing (really a question) that I just realized  It seems like we are planning to change what body each interceptor in receives in #269 This seems like a breaking change for a user using multiple interceptors so my question is should we attempt #269 as part of 0.2?
> While this is definite scope creep, it might be worth reducing the number of breaking changes?

Yeah I think this makes sense. This is likely the expected behavior anyway, I just didn't want to do everything in this one change.

**Pull request merged**